sources/bing_webmaster/__init__.py

```diff
@@ -0,0 +1,92 @@
+) -> Sequence[DltResource]:
+    """
+    A dlt source for the Bing Webmaster api.
+    It groups resources for the APIs which return organic search traffic statistics
+    Args:
+        site_urls: List[str]: A list of site_urls, e.g, ["dlthub.com", "dlthub.de"]. Use this if you need the weekly traffic per site_url and page
+        site_url_pages: Iterable[DictStrStr]: A list of pairs of site_url and page. Use this if you need the weekly traffic per site_url, page, and query
+    Returns:
+        Sequence[DltResource]: A sequence of resources that can be selected from including page_stats and page_query_stats.
+    """
+    return (
+        page_stats(site_urls),
+        page_query_stats(site_url_pages),
+    )
+
+
+@dlt.resource(
+    write_disposition="merge",
+    merge_key=("date", "page", "site_url"),
+    primary_key=("date", "page", "site_url"),
+    table_name="bing_page_stats",
+)
+def page_stats(
+    site_urls: List[str], bing_webmaster_api_key: str = dlt.secrets.value
+) -> Iterator[Iterator[DictStrAny]]:
+    """
+    Get detailed traffic statistics for top pages belonging to a site_url
+    API documentation:
+    https://learn.microsoft.com/en-us/dotnet/api/microsoft.bing.webmaster.api.interfaces.iwebmasterapi.getpagestats
+    Args:
+        site_urls (List[str]): List of player usernames to retrieve archives for.
+    Yields:
+        Iterator[Dict[str, Any]]: An iterator over list of organic traffic statistics.
+    """
+    api_path = "GetPageStats"
+    for site_url in site_urls:
+        params = {"siteUrl": site_url, "apikey": bing_webmaster_api_key}
+        print(f"Fetching for site_url: {site_url}")
+        response = get_stats_with_retry(api_path, params)
+        if len(response) > 0:
+            yield parse_response(response, site_url)
+
+
+@dlt.resource(
+    write_disposition="merge",
+    merge_key=("date", "page", "site_url", "query"),
+    primary_key=("date", "page", "site_url", "query"),
+    table_name="bing_page_query_stats",
+)
+def page_query_stats(
+    site_url_pages: Iterable[DictStrStr],
+    bing_webmaster_api_key: str = dlt.secrets.value,
+) -> Iterator[Iterator[DictStrAny]]:
+    """
+    Yields weekly statistics and queries for each pair of page and site_url.
+    Contains the entire available history. Thus, we recommend to use write_disposition="merge"
+    API documentation:
+    https://learn.microsoft.com/en-us/dotnet/api/microsoft.bing.webmaster.api.interfaces.iwebmasterapi.getpagequerystats
+
+    Args:
+        site_url_page (Iterable[DictStrStr]): Iterable of site_url and pages to retrieve statistics for. Can be result of a SQL query, a parsed sitemap, etc.
+    Yields:
+        Iterator[Dict[str, Any]]: An iterator over list of organic traffic statistics.
+    """
+    api_path = "GetPageQueryStats"
+    for record in site_url_pages:
+        time.sleep(0.5)  # this avoids rate limit observed after dozens of requests
```

Comment:
don't you want to handle the rate limit error instead of sleeping on every iteration?
here is an example: https://github.com/dlt-hub/verified-sources/blob/master/sources/workable/workable_client.py#L41

also there is no pagination?

Reply:
It is some weird MSFT API which does this kind of rate limiting, but if @willi-mueller has some time could we research if they have better alternative to handle rate limited endpoints?

Reply:
Unfortunately, they just give an unspecific HTTP 400 authentication error and the API access is blocked for minutes. Thus, I would keep the sleeping – otherwise, I observed that after around 10min of hammering the API with requests (e.g. while iterating over a sitemap) our pipeline would have to wait a couple of minutes until it's unblocked. I haven't done systematic experiments to determine what would be a backoff time after getting this error.

In this way, I've been able to consistently retrieve data for 2100 pages (i.e. 2100 requests).

Good idea to experimentally research the specific rate limit though, that would be cleaner than putting a sleep timer.

Reply:
crazy API :D

Reply:
Indeed, some interesting design decisions. Those were the times before they had access to ChatGPT 🙃 

![Ballmer I Love This Company](https://i.makeagif.com/media/7-13-2023/nTm-ut.gif)